### PR TITLE
Create "promote public proposal" extrinsic 

### PR DESCRIPTION
**Pull request by Doordashcon**
This PR attempts to create an extrinsic allowing members of council promote a public proposal

### CODE REVIEW

File: frame/democracy/src/lib.rs [hunk 1]
```diff
@@ -659,6 +677,25 @@ pub mod pallet {
 			Ok(())
 		}
 
+        #[pallet::weight(T::WeightInfo::promote())]
+        pub fn promote(
+            origin: OriginFor<T>,
+            proposal_hash: T::Hash,
+        ) -> DispatchResult {
```
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:684]: When adding a new extrinsic to a pallet, it is important that you place it at the bottom of the list of extrinsics, else you will modify the extrinsic ordering, and break the transaction format.

File: frame/democracy/src/lib.rs [hunk 2]
```diff
@@ -382,7 +385,18 @@ pub mod pallet {
...
-	/// Those who have locked a deposit.
+	/// Those who have promoted a particular public proposal.
+    #[pallet::storage]
+    #[pallet::getter(fn promoted)]
+    pub type Promoted<T: Config> = 
+        StorageMap<
+        _,
+        Twox64Concat, 
+        T::Hash, 
+        Vec<T::AccountId>
```
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:396]: should be bounded

**Comment by shawntabrizi** [comment]
Does this address a specific open issue? Can you link to it or provide a specification and user story for these changes?

File: frame/democracy/src/lib.rs [hunk 3]
```diff
@@ -659,6 +677,25 @@ pub mod pallet {
...
+            let public_props = PublicProps::<T>::get();
+            public_props.iter().position(|p| p.1 == proposal_hash).ok_or_else(|| Error::<T>::ProposalMissing)?;
+
+            let mut promoted_by = <Promoted<T>>::get(&proposal_hash).unwrap_or_else(Vec::new);
+            let insert_positions = promoted_by.binary_search(&who).err().ok_or(Error::<T>::AlreadyPromoted)?;
+            promoted_by.insert(insert_positions, who.clone());
+            <Promoted<T>>::insert(&proposal_hash, promoted_by);
+            Self::deposit_event(Event::<T>::PromotedBy { proposal_hash, who });
+
+            Ok(())
```
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:696]: What happens after something is promoted?

How does this storage item you are creating get cleaned up?

**Comment by Doordashcon** [comment]
> Does this address a specific open issue? Can you link to it or provide a specification and user story for these changes?

@joshua-mir

**Comment by joshua-mir** [comment]
sorry, have notifications muted. 

no open issue, just a discussion I had with Seun during his parachain recovery attempt - a proposal was made to the public democracy queue which precluded TC from being able to fast-track it, the idea behind this is it would be nice if it were possible for the queue that has "more privileges" to "adopt" proposals from other queues instead of having to submit a modified proposal in scenarios that are time-critical - might still be worth implementing here even if there is a democracy overhaul that would address the problem in another way.

File: frame/democracy/src/lib.rs (hunk 1 above)
Comment (Doordashcon) [reply frame/democracy/src/lib.rs]: This has been resolved

File: frame/democracy/src/lib.rs (hunk 2 above)
Comment (Doordashcon) [reply frame/democracy/src/lib.rs]: Changed to bounded made

File: frame/democracy/src/lib.rs (hunk 3 above)
Comment (Doordashcon) [reply frame/democracy/src/lib.rs]: It is cleaned up when a proposal gets moved to the external queue and should also be cleaned when canceled and blacklisted

**Comment by shawntabrizi** [comment]
I dont think a PR will be accepted here without a specification, discussion, and agreement about the approach.

**Pull request closed without merging**